gson/src/main/java/com/google/gson/internal/bind/ReflectiveTypeAdapterFactory.java

```diff
@@ -215,6 +228,25 @@ protected BoundField(String name, boolean serialized, boolean deserialized) {
         in.beginObject();
         while (in.hasNext()) {
           String name = in.nextName();
+          if (this.defaultTyping && TYPE_FIELD_NAME.equals(name)) {
+            String className = in.nextString();
+            if (PROPERTIES_FIELD_NAME.equals(in.nextName())) {
+              try {
+                Class<?> clazz = Class.forName(className, false, this.getClass().getClassLoader());
+                if (!instance.getClass().equals(clazz) && instance.getClass().isAssignableFrom(clazz)) {
+                  return (T) context.getAdapter(clazz).read(in);
+                } else {
+                  in.beginObject();
+                  continue;
+                }
+              } catch (ClassNotFoundException e) {
```

Comment:
Maybe it would be cleaner if this `try-catch` only contained the `Class.forName`, e.g.:
```
Class<?> clazz;
try {
  clazz = Class.forName(...);
} catch (ClassNotFoundException e) {
  throw ...
}
```